### PR TITLE
[config-engine] minigraph.py refactoring

### PR DESCRIPTION
**Pull request by taoyl-ms**
There is inconsistency among the interface of different field in current minigraph.py, and that is causing confusion and leads to misconfiguration in some of our j2 templates. This PR makes the following modification to the minigraph.py output:

- Consolidate the port/interface output. Now we have `minigraph_portchannels` and `minigraph_vlans` which are **dictionaries** with port channel name / vlan name as key, and portchannel members / vlan members and vlan ids as value. The configuration that needs to be generated by iterating though all portchannels / vlans should use these fields. 
- On the other hand, we have `minigraph_portchannel_interfaces` and `minigraph_vlan_interfaces` for the ip address configured on portchannels and vlans. Those fields are formatted in **lists** and have the similar format with `minigraph_interfaces`. Those fields should be used when needs to iterate though all ip addresses / networks.
- This PR also fixes several templates where they iterate through the wrong field and have potential issue when there are multiple addresses configured on a single port.
- To keep it consistent, front panel ports now also have a `minigraph_ports` dictionary, together with the existing `minigraph_interfaces` list. The current values of this dictionary are port names and alias, and we can consider extending it with front panel indices.
- In the previous version, `name` and `alias` in `minigraph_interfaces` are opposite from which in port_config.ini. This is fixed in the new `minigraph_ports`. (`name` - sonic name; `alias` - vendor-specific name)
- `minigraph_interfaces` will no longer have `name` and `alias`, but only a `attachto` where port sonic name will be put into.
- As `alias` information is included in `minigraph_ports`, `alias_map` field is no longer needed.

### CODE REVIEW

**Comment by taoyl-ms** [comment]
Sample outputs after this PR:
- minigraph_ports:
`'Ethernet0': {'alias': 'fortyGigE0/0', 'name': 'Ethernet0'}, 'Ethernet4': {'alias': 'fortyGigE0/4', 'name': 'Ethernet4'}}`

- minigraph_vlans:
`{'Vlan1000': {'name': 'Vlan1000', 'members': ['Ethernet0', 'Ethernet4']}}`

- minigraph_portchannels:
`{'PortChannel01': {'name': 'PortChannel13', 'members': ['Ethernet8']}}`

- minigraph_interfaces / minigraph_portchannel_interfaces / minigraph_vlan_interfaces (in the same format):
`[{'subnet': IPv4Network('10.0.0.56/31'), 'peer_addr': IPv4Address('10.0.0.57'), 'addr': IPv4Address('10.0.0.56'), 'mask': IPv4Address('255.255.255.254'), 'attachto': 'PortChannel01', 'prefixlen': 31}, {'subnet': IPv6Network('fc00::70/126'), 'peer_addr': IPv6Address('fc00::72'), 'addr': IPv6Address('fc00::71'), 'mask': '126', 'attachto': 'PortChannel01', 'prefixlen': 126}]`

**Comment by stcheng** [comment]
Based on the sample output, when will the SONiC alias used? when will the 'fortyGigE0/4' used? Will there be any replacement in the output?

**Comment by stcheng** [comment]
Could you also add some basic tests to ensure some output?

File: files/image_config/interfaces/interfaces.j2 [hunk 1]
```diff
@@ -43,58 +43,61 @@ iface eth0 inet dhcp
...
 {% for interface in minigraph_interfaces %}
-auto {{ interface['alias'] }}
-allow-hotplug {{ interface['alias'] }}
-iface {{ interface['alias'] }} {{ 'inet' if interface['addr'] | ipv4 else 'inet6' }} static
+iface {{ interface['attachto'] }} {{ 'inet' if interface['addr'] | ipv4 else 'inet6' }} static
     address {{ interface['addr'] }}
     netmask {{ interface['mask'] }}
-#
 {% endfor %}
-{% for vlan_interface in minigraph_vlan_interfaces|unique_name %}
```
Comment (stcheng) [inline files/image_config/interfaces/interfaces.j2]: could you also remove the unique_name filter in sonic-cfggen?

File: files/image_config/interfaces/interfaces.j2 [hunk 2]
```diff
@@ -43,58 +43,61 @@ iface eth0 inet dhcp
...
 {% endfor %}
 {% endfor %}
 {% endblock front_panel_interfaces %}
 {% block vlan_interfaces %}
+
+# Vlan interfaces
+{% for vlan in minigraph_vlans.keys()|sort %}
+auto {{ vlan }}
+{% endfor %}
+
```
Comment (stcheng) [inline files/image_config/interfaces/interfaces.j2]: do not have an empty line between the 'auto' line and the 'iface' line

File: files/image_config/interfaces/interfaces.j2 [hunk 3]
```diff
@@ -43,58 +43,61 @@ iface eth0 inet dhcp
...
 {% for vlan_interface in minigraph_vlan_interfaces %}
-auto {{ vlan_interface['name'] }}
-iface {{ vlan_interface['name'] }} {{ 'inet' if vlan_interface['addr'] | ipv4 else 'inet6' }} static
+iface {{ vlan_interface['attachto'] }} {{ 'inet' if vlan_interface['addr'] | ipv4 else 'inet6' }} static
     bridge_ports none
```
Comment (stcheng) [inline files/image_config/interfaces/interfaces.j2]: move it above

File: files/image_config/interfaces/interfaces.j2 [hunk 4]
```diff
@@ -43,58 +43,61 @@ iface eth0 inet dhcp
...
-#
 {% endfor %}
 {% endblock vlan_interfaces %}
 {% block pc_interfaces %}
+
+# Portchannel settings
+{% for pc in minigraph_portchannels.keys()|sort %}
+auto {{ pc }}
+allow-hotplug {{ pc }}
+{% endfor %}
```
Comment (stcheng) [inline files/image_config/interfaces/interfaces.j2]: i don't know if it is good to separate the iface stanza with the auto stanzas.

File: files/image_config/interfaces/interfaces.j2 (hunk 2 above)
Comment (stcheng) [inline files/image_config/interfaces/interfaces.j2]: it turns out to be that bridge_ports should follow iface line. thanks @taoyl-ms . let's stick auto stanza with iface stanza together for better readability.

**Comment by lguohan** [comment]
does this pass the sonic-cfggen unit test? Is it running?

**Comment by taoyl-ms** [comment]
The only output where 'fortyGigE0/4' will present is in the "alias" field of minigraph_ports. All the other fields use SONiC name.

---
In reply to: [289641473](https://github.com/Azure/sonic-buildimage/pull/448#issuecomment-289641473) [](ancestors = 289641473)

**Comment by taoyl-ms** [comment]
It seems the test is not running automatically after we moved from deb to whl. Just edited slave.mk to reenable it.

---
In reply to: [290018772](https://github.com/Azure/sonic-buildimage/pull/448#issuecomment-290018772) [](ancestors = 290018772)

File: files/image_config/interfaces/interfaces.j2 (hunk 1 above)
Comment (taoyl-ms) [inline files/image_config/interfaces/interfaces.j2]: I feel it's OK to keep it?

---
In reply to: [108321887](https://github.com/Azure/sonic-buildimage/pull/448#discussion_r108321887) [](ancestors = 108321887)

**Comment by taoyl-ms** [comment]
Upload a sample `/etc/network/interfaces` file for T0 after this refactor.
[interfaces.txt](https://github.com/Azure/sonic-buildimage/files/880470/interfaces.txt)

**Comment by taoyl-ms** [comment]
All CR comments resolved. Please help review.

**Comment by taoyl-ms** [comment]
#409

**Comment by stcheng** [comment]
`error: [Errno 2] No such file or directory: 'build/bdist.linux-x86_64/wheel/openconfig_acl.py'`
do you know why this error appears?

**Comment by stcheng** [comment]
resolve the conflicts?

**Pull request merged**